### PR TITLE
Guard against (de)serializing SerializationInfo instances

### PR DESCRIPTION
**Pull request by layomia**
Also prevents the serializer from calling parameterized ctors with signature `.ctor(SerializationInfo, StreamingContext)`.

FYI @GrabYourPitchforks, @eiriktsarpalis

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 1]
```diff
@@ -21,7 +21,9 @@ public sealed partial class JsonSerializerOptions
         // The global list of built-in converters that override CanConvert().
         private static readonly JsonConverter[] s_defaultFactoryConverters = new JsonConverter[]
         {
-            // Nullable converter should always be first since it forwards to any nullable type.
+            // Check for black-listed types.
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:24]: This is going to hit policheck.

```suggestion
            // Check for disallowed types.
```

**Comment by GrabYourPitchforks** [comment]
Given time I imagine we'll also want to deny the types `Assembly`, `MethodInfo`, and a whole slew of reflection-related things. Not really needed for this PR, but it would be interesting to think about whether the solution here might scale to capturing all of those. At the very least, in this theoretical future world, we'd want to update the exception message to avoid suggesting that every such type might lead to security problems.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverterFactory.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class UnsupportedTypeConverterFactory : JsonConverterFactory
+    {
+        public override bool CanConvert(Type type)
+        {
+            // If a value type is added, also add a test that
+            // shows NSE is thrown when Nullable<T> is deserialized.
+            return type == typeof(Type) || type == typeof(SerializationInfo);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverterFactory.cs:15]: What is the problem with these types?  I'd like to see a comment for everything we're explicitly blocking to say why.

**Review by eiriktsarpalis — COMMENTED** [review]
Is this getting backported for 5.0? If not, wouldn't this be a 6.0 breaking change for the hypothetical customer code that relies on `SerializationInfo`?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverter.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class UnsupportedTypeConverter<T> : JsonConverter<T>
+    {
+        private static readonly string s_exceptionMessage = SR.Format(SR.SerializeTypeInstanceNotSupported, typeof(T).FullName);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverter.cs:8]: Do we cache resource strings elsewhere? IMO we should just re-create the instance every time and save a bit of private memory.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverterFactory.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UnsupportedTypeConverterFactory.cs:15]: I've added comments in the source.

**Comment by layomia** [comment]
> it would be interesting to think about whether the solution here might scale to capturing all of those. At the very least, in this theoretical future world, we'd want to update the exception message to avoid suggesting that every such type might lead to security problems.

wrt. adapting the error message appropriately, we could add a new internal `TypeDisallowReason` enum (or something similar) which would be a parameter for a ctor on `UnsupportedTypeConverter` which would inform what message to spit out. Hopefully any further scalability issue can be worked out.

**Comment by layomia** [comment]
> Is this getting backported for 5.0? If not, wouldn't this be a 6.0 breaking change for the hypothetical customer code that relies on SerializationInfo?

This change won't be a breaking change in 6.0, so we don't need to port it to 5.0. The `Type` and `SerializationInfo` types are already not (de)serializable, due to `Type` already being explicitly disallowed (both ctors on `SerializationInfo` take a `Type`, and there is serializable `Type` member on `SerializationInfo`). This change only makes the exception message clearer when deserializing `SerializationInfo` and when disallowing `.ctor(SerializationInfo, StreamingContext)` from being invoked.

**Review by steveharter — APPROVED** [review]

**Comment by layomia** [comment]
Test failures appear unrelated - 2 timeouts:

`runtime (Libraries Test Run release coreclr Linux x64 Debug) Cancelled after 150m`

`runtime (Libraries Test Run release mono Linux x64 Debug) Cancelled after 149m`

**Pull request merged**